### ISSUE TITLE
BUG with autoreload: modify readonly __closure__ attribute

### ISSUE THREAD

**Issue by phihung** [ipython/ipython#14525]
The [update_function](https://github.com/ipython/ipython/blob/6c00ceacfaacfe2766e4170a11fc95621545de4c/IPython/extensions/autoreload.py#L309) in the `autoreload` module tries to copy the `__closure__` attribute from the new version of the function to the old one (presumably for reloading decorated functions).

However, `__closure__` is a read-only attribute, so the copy never actually takes place. The error is caught and ignored.

This may be the cause of the long-standing issue "autoreload does not work with decorators": [link](https://github.com/ipython/ipython/issues/2543) and [link](https://github.com/ipython/ipython/issues/11099).

The fix should be quite simple.
```python
old_closure, new_closure = old.__closure__, new.__closure__
if len(old_closure) != len(new_closure):
    # may need more check for variable names
    return
for o, n in zip(old_closure, new_closure):
    o.cell_contents = n.cell_contents
```

Have I just found a 16-year-old bug? : )

**Comment by krassowski** [comment]
Would it be solved by https://github.com/ipython/ipython/pull/14500?

**Comment by phihung** [comment]
Yes.
I did the tests. The `DeduperReloader` implementation correctly updates __closure__ attribute.
But as I understand, there will be two modes of reloading: the new DeduperReloader and the old one ?
If that is the case, the old one still needs to be fixed

**Comment by smacke** [comment]
That's correct that deduperreload could fall back to the original algorithm if it can't isolate module changes to individual functions.

> The fix should be quite simple.

Unfortunately the proposed fix only works if the old and new closures have the same length. Deduperreload works around this by doing some magic to patch the readonly `__closure__` attribute at the level of the C struct; we could potentially do the same thing in the original autoreload algorithm as well.

> Have I just found a 16-year-old bug? : )

It's hard to say whether it's a bug or a limitation -- I think autoreload is one of those "best-effort" things and it's pretty easy to hit a case with undefined behavior.
